fence/man/fence_node.8
.TH FENCE_NODE 8 2009-12-21 cluster cluster

.SH NAME
fence_node \- a utility to run fence agents

.SH SYNOPSIS
.B fence_node
[OPTIONS]
.I nodename

.SH DESCRIPTION
This utility runs a fence agent against
.IR nodename .
The agent and args are taken from the running cluster configuration based on
.BR cluster.conf (5).

.P
.B fence_node
is a wrapper around the libfence functions: fence_node() and unfence_node().
These libfence functions use libccs to read the node fencing configuration,
which means that corosync (with cman and ccs) must be running to use
.BR fence_node (8).

.P
The
.BR fenced (8)
daemon is the main user of libfence:fence_node(), and the
configuration details for that function are given in the
.BR fenced (8)
man page.

.SS Fencing vs. Unfencing

The main use for unfencing is with storage/SAN (non-power) agents.

.P
When using power-based fencing agents, the fencing action itself is
supposed to turn a node back on after first turning the power off (this happens
automatically with a "reboot" action, and needs to be configured
explicitly as "off" + "on" otherwise.)

.P
When using storage-based fencing agents, the fencing action is not allowed
to re-enable a node after disabling it.  Re-enabling a fenced node is only
safe once the node has been rebooted.  A natural way to re-enable a fenced
node's access to storage, is for that node to re-enable the access itself
during its startup process.  The cman init script calls fence_node -U
(nodename defaults to local nodename when unfencing).  Unfencing a node
without an <unfence> configuration (see below) is a no-op.

.P
The basic differences between fencing and unfencing:
.P
.BR Fencing
.IP 1. 3
libfence: fence_node(), command line: fence_node nodename
.IP 2. 3
Turns off or disables a node.
.IP 3. 3
Agents run with the default action of "off", "disable" or "reboot".
.IP 4. 3
Performed by a cluster node against another node that fails (by the fenced daemon).
.P
.BR Unfencing
.IP 1. 3
libfence: unfence_node(), command line: fence_node -U nodename
.IP 2. 3
Turns on or enables a node.
.IP 3. 3
Agents run with the explicit action of "on" or "enable".
.IP 4. 3
Performed by a cluster node "against" itself during startup (by the cman init script).

.SH OPTIONS
.TP
.B \-U
Unfence the node, default local node name.
.TP
.B \-v
Show fence agent results, \-vv to also show agent args.
.TP
.B \-h
Print a help message describing available options, then exit.
.TP
.B \-V
Print program version information, then exit.

.SH FILES

The Unfencing/unfence_node() configuration is very similar to the
Fencing/fence_node() configuration shown in
.BR fenced (8).
Unfencing is only performed for a node with an <unfence> section:

.nf
<clusternode name="node1" nodeid="1">
        <fence>
        </fence>
        <unfence>
        </unfence>
</clusternode>
.fi

The <unfence> section does not contain <method> sections like the <fence>
section does.  It contains <device> references directly, which mirror the
corresponding device sections for <fence>, with the notable addition of
the explicit action of "on" or "enable".  The same <fencedevice> is
referenced by both fence and unfence <device> lines, and the same per-node
args should be repeated.

.nf
<clusternode name="node1" nodeid="1">
        <fence>
        <method name="1">
        <device name="myswitch" foo="x"/>
        </method>
        </fence>

        <unfence>
        <device name="myswitch" foo="x" action="on"/>
        </unfence>
</clusternode>
.fi

.SH SEE ALSO
.BR fenced (8)



rgmanager/ChangeLog
2008-04-16 Fabio M. Di Nitto <fabbione at fabbione.net>
	* src/utils/clustat.c: Fix variable init

2008-03-04 Lon Hohberger <lhh at redhat.com>
	* src/resources/ASEHAagent.sh, Makefile: Add Sybase failover agent
	* src/resources/oracledb.sh, Makefile: Add Oracle 10g failover agent

2008-02-26 Lon Hohberger <lhh at redhat.com>
	* src/resources/ip.sh: Fix netmask handling in ip.sh
	* src/utils/clustat.c: Don't show estranged nodes if they're down
2008-02-01 Lon Hohberger <lhh at redhat.com>
	* src/daemons/rg_state.c, slang_event.c, groups.c, restart_counter.c,
	include/restart_counter.h, resgroup.h,
	src/resources/default_event_script.sl: Allow restart counters to 
	correctly work with central_processing. (#400211 / #431130)

2008-01-25 Lon Hohberger <lhh at redhat.com>
	* src/daemons/rg_thread.c: Fix case that broke 'clusvcadm -e <service>
	-n <node>' case #430220
	* src/daemons/rg_state.c: Don't enable migrate-to-self (#430272)

2008-01-02 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/resources/ocf-shellfuncs: Add interpreter.

2007-12-30 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/daemons/rg_state.c, include/resgroup.h, src/clulib/alloc.c:
	  Fix building when -DDEBUG is defined.

2007-12-24 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/clulib/vft.c: Change ifdef to fix build on parisc.

2007-12-19 Lon Hohberger <lhh at redhat.com>
	* include/res-ocf.h: Fix typo
	* src/daemons/restree.c: Export reference counts to resource agents.
	* src/resources/clusterfs.sh: Use reference counts to ensure we
	don't unmount a GFS file system if it's in use by multiple services
	on the same node.  #254111
	* src/resources/service.sh, default_event_script.sl: Add support for
	"soft" dependencies if central_processing is enabled

2007-12-14 Lon Hohberger <lhh at redhat.com>
	(Misc central_events fixes)
	* Add return code for inability to run because of exclusive 
	tag
	* Only initialize state once
	* Fix master-node failover.
	* Correctly return node after enable if request was forwarded

2007-12-12 Lon Hohberger <lhh at redhat.com>
	* Misc changes; add missing ds.h
	* src/resources/default*.sl: Make clusvcadm -r go to a different
	node.  Don't notice() that we're starting a service if it's
	disabled.
	* src/daemons/groups.c: Make clustat not interfere with startup
	when central processing is enabled
	* src/daemons/slang_event.c: Make clusvcadm correctly report
	service owner after relocate/restart/enable requests.

2007-11-30 Lon Hohberger <lhh at redhat.com>
	* Commit RIND / S-Lang script engine [untested]

[RHEL5 merged ChangeLog Entries]
2007-11-30 Lon Hohberger <lhh at redhat.com>
	* src/resources/clusterfs.sh: Retry mount up to 3 times to avoid
	race condition during another process mounting a GFS volume
	* src/resources/vm.sh, service.sh: Add defaults for values.
	Make vm.sh work with more service attrs (max restarts)
	* src/utils/clustat.c: Make output of clustat terminal-width 
	dependent

2007-11-26 Lon Hohberger <lhh at redhat.com>
	* include/reslist.h: Add restart counters to resource node structure
	(intended for top-level resources, i.e. services, vms...)
	* include/restart_counter.h: Add header file for restart counter
	* src/daemons/Makefile: Fix build to include restart counters
	* src/daemons/restart_counter.c: Implement restart counters #247139
	* src/daemons/fo_domain.c, groups.c, restart_counter.c, resrules.c,
	restree.c, test.c: Glue for restart counters.
	* src/daemons/reslist.c: Glue for restart counters.  Make expand_time
	parser more robust to allow things like '1h30m' as a time value.
	* src/daemons/main.c: Mark quorum disk offline in the correct
	place to avoid extraneous log messages
	* src/daemons/rg_state.c: Allow marking service as stopped if
	stuck in recover state.  Make service which failed to start
	go to stopped state.  Glue for restart counters.
	* src/resources/service.sh, vm.sh: Add parameters for restart
	counters #247139

2007-11-14 Lon Hohberger <lhh at redhat.com>
	* src/utils/clulog.c: Make clulog honor rgmanager log levels
	(#289501)
	* src/clulib/vft.c: Fix #303981 - crash on rgmanager restart in some
	cases
	* man/clusvcadm.8: Remove references to clushutdown from man page;
	resolves #324151
	* src/resources/netfs.sh: Apply patch from Marco Ceci to fix #358161
	* src/resources/vm.sh: Make default migration policy live instead
	of pause for Xen virtual machines.  Also make it configurable instead
	of static.  Resolves #345871

2007-11-13 Lon Hohberger <lhh at redhat.com>
	* src/resources/clusterfs.sh: Add support for self_fence operation
	to clusterfs resource agent
	* src/resources/service.sh: Add default values to service.sh

2007-10-26 Lon Hohberger <lhh at redhat.com>
	* src/daemons/main.c, src/utils/clustat.c, clusvcadm.c:
	Fix #354391

2007-09-25 Lon Hohberger <lhh at redhat.com>
	* src/daemons/restree.c: Apply patch to fix side case re: 229650
	Patch from Simone Gotti. Resolves: #229650
[End RHEL5 merged changes]

2007-08-30 Lon Hohberger <lhh at redhat.com>
	* src/daemons/restree.c, rg_state.c: Fix tree-restart bug
	This is another part of #229650

2007-08-22 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* include/list.h: Fix build with gcc-4.2.

2007-08-15 Lon Hohberger <lhh at redhat.com>
	* src/utils/clustat.c: Remove uninitialized var & fix logic /
	spotted by Ryan McCabe

2007-08-09  Fabio M. Di Nitto <fabbione at ubuntu.com>
	* rgmanager/src/clulib/alloc.c: Undefine DEBUG when building on
	PARISC. __builtin_address needs the same love as IA64.

2007-08-02 Lon Hohberger <lhh at redhat.com>
	* general: More fixes around #248727
	* include/reslist.h, src/daemons/restree.c: Make last-value be
	returned or resources which have been checked recently
	* src/daemons/groups.c: Make VMs use migrate semantics instead of
	relocate semantics when employing failover domain rules
	* src/daemons/nodeevent.c: Fix VMs ending up on wrong nodes when
	simultaneous boot occurs
	* src/daemons/rg_forward.c: Fix erroneous timeout

fence/fence_node/fence_node.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdint.h>
#include <string.h>
#include <liblogthread.h>
#include <libcman.h>

#include "libfence.h"
#include "libfenced.h"
#include "copyright.cf"

static char *prog_name;
static char our_name[CMAN_MAX_NODENAME_LEN+1];
static int verbose;
static int unfence;

#define FL_SIZE 32
static struct fence_log flog[FL_SIZE];
static int flog_count;
static const char *action = "fence";

#define OPTION_STRING "UvhV"

#define die(fmt, args...) \
do \
{ \
  fprintf(stderr, "%s: ", prog_name); \
  fprintf(stderr, fmt "\n", ##args); \
  exit(EXIT_FAILURE); \
} \
while (0)

static void print_usage(void)
{
	printf("Usage:\n");
	printf("\n");
	printf("%s [options] node_name\n", prog_name);
	printf("\n");
	printf("Options:\n");
	printf("\n");
	printf("  -U    Unfence the node, default local node name\n");
	printf("  -v    Show fence agent results, -vv for agent args\n");
	printf("  -h    Print this help, then exit\n");
	printf("  -V    Print program version information, then exit\n");
	printf("\n");
}

static int setup_cman(void)
{
	cman_handle_t ch;
	cman_node_t node;
	int active = 0;
	int rv;

	ch = cman_init(NULL);
	if (!ch)
		return -1;

 retry_active:
	rv = cman_is_active(ch);
	if (!rv) {
		if (active++ < 2) {
			sleep(1);
			goto retry_active;
		}
		cman_finish(ch);
		return -1;
	}

	memset(&node, 0, sizeof(node));
	rv = cman_get_node(ch, CMAN_NODEID_US, &node);
	if (rv < 0) {
		cman_finish(ch);
		return -1;
	}

	memset(our_name, 0, sizeof(our_name));
	strncpy(our_name, node.cn_name, CMAN_MAX_NODENAME_LEN);
	cman_finish(ch);
	return 0;
}

static const char *fe_str(int r)
{
	switch (r) {
	case FE_AGENT_SUCCESS:
		return "success";
	case FE_AGENT_ERROR:
		return "error from agent";
	case FE_AGENT_FORK:
		return "error from fork";
	case FE_NO_CONFIG:
		return "error from ccs";
	case FE_NO_METHOD:
		return "error no method";
	case FE_NO_DEVICE:
		return "error no device";
	case FE_READ_AGENT:
		return "error config agent";
	case FE_READ_ARGS:
		return "error config args";
	case FE_READ_METHOD:
		return "error config method";
	case FE_READ_DEVICE:
		return "error config device";
	default:
		return "error unknown";
	}
}

int main(int argc, char *argv[])
{
	char *victim = NULL;
	int cont = 1, optchar, error, rv, i, c;

	prog_name = argv[0];

	while (cont) {
		optchar = getopt(argc, argv, OPTION_STRING);

		switch (optchar) {

		case 'U':
			unfence = 1;
			action = "unfence";
			break;

		case 'v':
			verbose++;
			break;

		case 'h':
			print_usage();
			exit(EXIT_SUCCESS);
			break;

		case 'V':
			printf("%s %s (built %s %s)\n", prog_name,
				RELEASE_VERSION, __DATE__, __TIME__);
			printf("%s\n", REDHAT_COPYRIGHT);
			exit(EXIT_SUCCESS);
			break;

		case ':':
		case '?':
			fprintf(stderr, "Please use '-h' for usage.\n");
			exit(EXIT_FAILURE);
			break;

		case EOF:
			cont = 0;
			break;

		default:
			die("unknown option: %c", optchar);
			break;
		};
	}

	while (optind < argc) {
		if (victim)
			die("unknown option %s", argv[optind]);
		victim = argv[optind];
		optind++;
	}

	if (!victim && !unfence)
		die("no node name specified");

	error = setup_cman();
	if (error)
		die("cannot connect to cman");

	if (!victim && unfence)
		victim = our_name;

	memset(&flog, 0, sizeof(flog));
	flog_count = 0;

	if (unfence)
		error = unfence_node(victim, flog, FL_SIZE, &flog_count);
	else
		error = fence_node(victim, flog, FL_SIZE, &flog_count);

	if (!verbose)
		goto skip;

	if (flog_count > FL_SIZE) {
		fprintf(stderr, "%s_node log overflow %d", action, flog_count);
		flog_count = FL_SIZE;
	}

	for (i = 0; i < flog_count; i++) {
		fprintf(stderr, "%s %s dev %d.%d agent %s result: %s\n",
			action, victim, flog[i].method_num, flog[i].device_num,
			flog[i].agent_name[0] ? flog[i].agent_name : "none",
			fe_str(flog[i].error));

		if (verbose < 2)
			continue;

		for (c = 0; c < strlen(flog[i].agent_args); c++) {
			if (flog[i].agent_args[c] == '\n')
				flog[i].agent_args[c] = ' ';
		}
		fprintf(stderr, "agent args: %s\n", flog[i].agent_args);
	}

 skip:
	logt_init("fence_node", LOG_MODE_OUTPUT_SYSLOG, SYSLOGFACILITY,
		  SYSLOGLEVEL, 0, NULL);

	if (unfence) {
		if (error == -2) {
			fprintf(stderr, "unfence %s undefined\n", victim);
			rv = 2;
		} else if (error) {
			fprintf(stderr, "unfence %s failed\n", victim);
			logt_print(LOG_ERR, "unfence %s failed\n", victim);
			rv = EXIT_FAILURE;
		} else {
			fprintf(stderr, "unfence %s success\n", victim);
			logt_print(LOG_ERR, "unfence %s success\n", victim);
			rv = EXIT_SUCCESS;
		}
	} else {
		if (error == -2) {
			fprintf(stderr, "fence %s undefined\n", victim);
			logt_print(LOG_ERR, "fence %s undefined\n", victim);
			rv = 2;
		} else if (error) {
			fprintf(stderr, "fence %s failed\n", victim);
			logt_print(LOG_ERR, "fence %s failed\n", victim);
			rv = EXIT_FAILURE;
		} else {
			fprintf(stderr, "fence %s success\n", victim);
			logt_print(LOG_ERR, "fence %s success\n", victim);
			rv = EXIT_SUCCESS;

			/* Tell fenced what we've done so that it can avoid
			   fencing this node again if the fence_node() rebooted
			   it. */
			fenced_external(victim);
		}
	}

	logt_exit();
	exit(rv);
}

